source/common/event/dispatcher_impl.cc

```diff
@@ -155,28 +153,46 @@ DispatcherImpl::createClientConnection(Network::Address::InstanceConstSharedPtr
+// the caller will program some other data in it, like use_tcp_for_dns_lookups or dns_resolvers.
+// @param dns_resolver_config could be empty.
 Network::DnsResolverSharedPtr DispatcherImpl::createDnsResolver(
-    const std::vector<Network::Address::InstanceConstSharedPtr>& resolvers,
-    const envoy::config::core::v3::DnsResolverOptions& dns_resolver_options) {
+    envoy::config::core::v3::DnsResolutionConfig& dns_resolution_config,
+    envoy::config::core::v3::TypedExtensionConfig& dns_resolver_config) {
+
   ASSERT(isThreadSafe());
-#ifdef __APPLE__
-  static bool use_apple_api_for_dns_lookups =
+
+  Network::DnsResolverFactory* dns_resolver_factory;
+
+  if (dns_resolver_config.typed_config().type_url().empty()) {
+    // If the typed dns resolver configuration is missing, synthetic a dns_resolver_config
+    // based on the run time use_apple_api_for_dns_lookups flag. Then using
+    // the dns_resolver_config to retrieve dns_resolver_factory.
+
+    static bool use_apple_api_for_dns_lookups =
       Runtime::runtimeFeatureEnabled("envoy.restart_features.use_apple_api_for_dns_lookups");
-  if (use_apple_api_for_dns_lookups) {
-    RELEASE_ASSERT(
-        resolvers.empty(),
-        "defining custom resolvers is not possible when using Apple APIs for DNS resolution. "
-        "Apple's API only allows overriding DNS resolvers via system settings. Delete resolvers "
-        "config or disable the envoy.restart_features.use_apple_api_for_dns_lookups runtime "
-        "feature.");
-    RELEASE_ASSERT(!dns_resolver_options.use_tcp_for_dns_lookups(),
-                   "using TCP for DNS lookups is not possible when using Apple APIs for DNS "
-                   "resolution. Apple' API only uses UDP for DNS resolution. Use UDP or disable "
-                   "the envoy.restart_features.use_apple_api_for_dns_lookups runtime feature.");
-    return std::make_shared<Network::AppleDnsResolverImpl>(*this, api_.rootScope());
+    if (use_apple_api_for_dns_lookups) {
+      // In Apple system, ignore the dns_resolution_config.
+      dns_resolver_config.mutable_typed_config()->PackFrom(Envoy::ProtobufWkt::Struct());
+    } else {
+      dns_resolver_config.mutable_typed_config()->PackFrom(dns_resolution_config);
+    }
   }
-#endif
-  return std::make_shared<Network::DnsResolverImpl>(*this, resolvers, dns_resolver_options);
+
+  // If a typed dns resolver is configured, derive the DNS resolver factory from the config.
+  dns_resolver_factory =
+      &Config::Utility::getAndCheckFactory<Network::DnsResolverFactory>(dns_resolver_config);
+  if (dns_resolver_factory != nullptr) {
+    ENVOY_LOG(info, "create DNS resolver type: {}", dns_resolver_config.name());
+    return dns_resolver_factory->createDnsResolverCb(*this,
+                                                     api_,
+                                                     dns_resolver_config);
+
+  }
+
+  return nullptr;
```

Comment:
Returning `null` here could be problematic. I suggest restructuring this function to be more like the previous in case a non-typed configuration is being passed, and if a typed-config is used then just call the factory with the config.
If no config is set, use the default one (keep previous behavior).

Reply:
Yes, no need to return a NULL pointer here. 

If getAndCheckFactory() can not find the corresponding factory (due to either the extension is not enabled in extension build file which leads to the corresponding dns resolver type is not registered, or the configured DNS typed extension doesn't match with any registered DNS factories ), it will throw an exception and abort anyway.